share/smack/top.py

```diff
@@ -899,85 +979,133 @@ def verify_bpl(args):
 
-        if args.error_file:
-            with open(args.error_file, 'w') as f:
-                f.write(error)
 
-        if not args.quiet:
-            print(error)
+def thread_verify_bpl(args, commands_to_add):
+    args.modular = False
 
-        if args.replay:
-            replay_error_trace(verifier_output, args)
-    print(result.message(args))
-    return result.return_code()
+    if 'verifier' in commands_to_add:
+        args.verifier = commands_to_add['verifier']
+    else:
+        args.verifier = 'corral'
+
+    if 'modular' in commands_to_add:
+        args.modular = commands_to_add['modular']
+
+    if 'verifier-options' in commands_to_add:
+        args.verifier_options = commands_to_add['verifier-options']
+
+    if 'solver' in commands_to_add:
+        args.solver = commands_to_add['solver']
+
+    if args.verifier == 'boogie' or args.modular:
+        command = invoke_boogie(args)
+
+    elif args.verifier == 'corral':
+        command = invoke_corral(args)
+
+    elif args.verifier == 'symbooglix':
+        command = invoke_symbooglix(args)
+
+    if args.verifier_options:
+        command += args.verifier_options.split()
+
+    return args, try_command(command, timeout=args.time_limit)
+
+
+def verify_bpl_portfolio(args):
+    if args.portfolio_config is None:
+        portfolio_config = {'thread1':
+                               {'verifier': 'corral',
+                                'verifier-options': ""},
+                            'thread2':
+                               {'verifier': 'corral',
+                                'verifier-options':
+                                    "/bopt:proverOpt:O:smt.arith.solver=2"},
+                            'thread3':
+                               {'verifier': 'corral',
+                                'verifier-options':
+                                    "/bopt:proverOpt:O:smt.qi.eager_threshold=100"},
+                            'thread4':
+                               {'verifier': 'corral',
+                                'verifier-options':
+                                   ("/bopt:proverOpt:O:smt.qi.eager_threshold=100"
+                                    " /bopt:proverOpt:O:smt.arith.solver=2")}}
+    else:
+        portfolio_config = yaml.load(open(args.portfolio_config, 'r'))
+
+    p = multiprocessing.Pool()
+    thread_args = copy.deepcopy(args)
+    results = {}  # map of process -> thread name
+
+    for thread in list(portfolio_config.keys()):
+        results[p.apply_async(thread_verify_bpl,
+                              args=(thread_args, portfolio_config[thread]),
+                              callback=get_result)] = thread
+
+    while True:  # sleep?
+        for result in list(results.keys()):
+            if result.ready():
+                p.terminate()
+                args, verifier_output = result.get()
+                verifier_output = process_verifier_output(args, verifier_output)
+
+                codes = {0 : 'verified',
```

Comment:
Why do you need these `codes`? Is this just for debugging?

Reply:
This was just to interpret the return code of trying a command with smack, but if the singular printout from process_verifier_output() is sufficient then I can remove these codes and the additional SMACK PORTFOLIO printout

Reply:
Let me know if I should delete these codes. Then the expected output would become something like
`SMACK found no errors with unroll bound 1.
SMACK PORTFOLIO: thread1 returned result code 0`
rather than
`SMACK found no errors with unroll bound 1.
SMACK PORTFOLIO: thread1 returned result code 0, corresponding to: verified`

Reply:
I think we can remove them since SMACK printout will already capture the status. We should say though which thread returned first. No need to add result code. So just something like "SMACK portfolio thread1 terminated".